### PR TITLE
felix/bpf: host does not require CTLB to access cluster IPs

### PR DESCRIPTION
**Pull request by tomastigera**
## Description

    felix/bpf: route service access from host via bpfnatin/out veth
    
    1 - program at the bfpnatout ingress to host is responsible for NAT and
        setting a via-nat-iface flag on conntrack.
    
    2 - program at wep should do "DSR" when the via-nat-iface flag is set on the
        conntrack entry if the backend is local. "DSR" means that the return
        traffic is not routed via bpfnatin/out.
    
    3 - program at hep should do the same for traffic from remote backends.
    
    Do not rely on iptables to fix up source when the backend pod is
    accessible via tunnel as that makes routing super complicated.
    
    We change only the source IP and we leave the port untouched as that
    would only collide if the host was also accessing the same pod directly
    (not via a service). If that happens, we we resolve the source port
    collision in the same way as we do for nodeports. That is, we pick a
    random port from a range that is not used for ephemeral ports. We can
    pick multiple times if there is a collision.

    felix/bpf: disable rp_filter at pods in bpf mode.
    
    BPF programs do the RPF check anyway, so this is not a hole. Needed for
    responding to traffic from local host to a services backed by a local
    pod. SNAT is done at the pod, kind of like DSR.

## Related issues/PRs

refs https://github.com/projectcalico/calico/issues/4509

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
ebpf: host does not require CTLB to access cluster IPs
```

### CODE REVIEW

File: felix/bpf-gpl/conntrack.h [hunk 1]
```diff
@@ -236,14 +243,25 @@ static CALI_BPF_INLINE int calico_ct_v4_create_nat_fwd(struct ct_create_ctx *ct_
...
-	__be32 ip_src = ct_ctx->src;
+	__be32 ip_src = ct_ctx->orig_src;
 	__be32 ip_dst = ct_ctx->orig_dst;
 	__u16 sport = ct_ctx->orig_sport;
 	__u16 dport = ct_ctx->orig_dport;
 
+	if (CALI_F_TO_HEP && !CALI_F_NAT_IF && sport != ct_ctx->sport &&
+			!(ct_ctx->skb->mark & (CALI_SKB_MARK_FROM_NAT_IFACE_OUT | CALI_SKB_MARK_SEEN))) {
+		/* This entry is being created because we have a source port
+		 * conflict on a connection from host. We did psnat so we mak
```
Comment (fasaxc) [inline felix/bpf-gpl/conntrack.h:254]: Typo: mak

File: felix/bpf-gpl/tc.c [hunk 2]
```diff
@@ -786,11 +801,26 @@ static CALI_BPF_INLINE struct fwd calico_tc_skb_accepted(struct cali_tc_ctx *ctx
...
 		state->ct_result.nat_sport = ct_ctx_nat.sport;
 		/* fall through as DNAT is now established */
 
+		if (CALI_F_TO_HOST && CALI_F_NAT_IF) {
+			struct cali_rt *r = cali_rt_lookup(state->post_nat_ip_dst);
+			if (r && cali_rt_flags_remote_workload(r->flags) && r->flags & CALI_RT_TUNNELED) {
+				CALI_DEBUG("remote wl %x tunneled\n", bpf_htonl(state->post_nat_ip_dst));
+				ct_ctx_nat.src = HOST_TUNNEL_IP;
+				/* This would be the place to set a new source port like
+				 * ct_ctx_nat.sport = 10101;
```
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:817]: Not sure what this comment means?  Whenever you use a phrase like "would be" it cries out for a second clause that explains why you're not doing it that way.

File: felix/bpf-gpl/tc.c [hunk 3]
```diff
@@ -1231,6 +1280,73 @@ int calico_tc_skb_send_icmp_replies(struct __sk_buff *skb)
...
+	switch (ct_result_rc(ctx.state->ct_result.rc)) {
+	case CALI_CT_ESTABLISHED:
+		/* Because we are on the "from host" patch, conntrack gives us
+		 * CALI_CT_ESTABLISHED only. Better to fix the corner case here than on
+		 * the generic path.
+		 */
+		ct_result_set_rc(ctx.state->ct_result.rc, CALI_CT_ESTABLISHED_DNAT);
+		/* fallthrough */
+	case CALI_CT_NEW:
+		/* There is a conflict, this is the first packet that conflictis. By
```
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:1326]: Typo: `conflictis`

File: felix/bpf-gpl/tc.c [hunk 4]
```diff
@@ -1231,6 +1280,73 @@ int calico_tc_skb_send_icmp_replies(struct __sk_buff *skb)
...
+		 * setting a NAT destination being the same as the original destination,
+		 * we trigger DNAT (void) which will conflict on the source port and will
```
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:1328]: What does `void` mean here?

File: felix/bpf/ut/snat_test.go [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+		Expect(res.dataOut).To(Equal(resPktBytes))
+
+		pktBytes = res.dataOut
+	})
+
+	skbMark = tcdefs.MarkSeen | tcdefs.MarkSeenFromNatIfaceOut | tcdefs.MarkSeenBypassSkipRPF
+
+	dumpCTMap(ctMap)
+
+	// Out via host iface. We shoul duse a L3 tunnel, but that is not supported
```
Comment (fasaxc) [inline felix/bpf/ut/snat_test.go:131]: Typo: shoul

File: felix/routetable/route_table.go [hunk 6]
```diff
@@ -120,7 +122,7 @@ func (t Target) RouteType() int {
 		return syscall.RTN_BLACKHOLE
 	case TargetTypeProhibit:
 		return syscall.RTN_PROHIBIT
-	case TargetTypeUnicast:
+	case TargetTypeUnicastGlobal, TargetTypeUnicastLinkLocal:
```
Comment (fasaxc) [inline felix/routetable/route_table.go:125]: D'oh, think we've already got this in Enterprise but with a different name.  Please can you upstream?

File: felix/bpf-gpl/fib.h [hunk 7]
```diff
@@ -212,7 +212,7 @@ static CALI_BPF_INLINE int forward_or_drop(struct cali_tc_ctx *ctx)
 		 * XXX those devices where we expect them before we even decap.
 		 */
 		if (CALI_F_FROM_HEP && state->tun_ip != 0 && ctx->fwd.mark != CALI_SKB_MARK_BYPASS_FWD) {
-			ctx->fwd.mark |= CALI_SKB_MARK_SKIP_RPF;
+			ctx->fwd.mark = CALI_SKB_MARK_SKIP_RPF;
```
Comment (fasaxc) [inline felix/bpf-gpl/fib.h:215]: Why this change, might be worth a comment to say which flags you're deliberately clobbering here

**Comment by fasaxc** [comment]
> That requires higher priority rule then the default local lookup

Think the PR description is out of date now?  I thought you avoided that by doing SNAT in BPF on collision?  (Or is that piece just not included in this PR...?)

**Review by fasaxc — CHANGES_REQUESTED** [review]
Did a first pass commit-by-commit (thanks; that made it a lot easier to grok!)  Still letting it sink in  but thought I'd release comments asap.

**Comment by tomastigera** [comment]
> > That requires higher priority rule then the default local lookup
> 
> Think the PR description is out of date now? I thought you avoided that by doing SNAT in BPF on collision? (Or is that piece just not included in this PR...?)

It does include it. The message was just a concatenation of the commit messages, I adjusted it to give a better picture.

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 8]
```diff
@@ -1349,13 +1388,84 @@ func (m *bpfEndpointManager) setAcceptLocal(iface string, val bool) error {
...
+	_, err = nl.LinkByName("bpfnatin")
+	if err != nil {
+		nat := &netlink.Veth{
+			LinkAttrs: netlink.LinkAttrs{
+				Name: "bpfnatin",
+			},
+			PeerName: "bpfnatout",
+		}
+		if err := nl.LinkAdd(nat); err != nil {
+			log.Fatal("Failed to add bpfnatin.")
```
[the remark this answers is not in the document]
Comment (tomastigera) [reply felix/dataplane/linux/bpf_ep_mgr.go:1434]: Ohhh yeah, residual of early coding, did not get to fix it up, will! Thanks for spotting this!

File: felix/bpf-gpl/tc.c (hunk 4 above)
Comment (tomastigera) [reply felix/bpf-gpl/tc.c:1328]: ~fake

File: felix/routetable/route_table.go (hunk 6 above)
Comment (tomastigera) [reply felix/routetable/route_table.go:125]: https://github.com/projectcalico/calico/pull/6082

File: felix/bpf-gpl/fib.h (hunk 7 above)
Comment (tomastigera) [reply felix/bpf-gpl/fib.h:215]: It was wrong to start with. It wasn't supposed to be ORed ever. NOw with RPF in BPF and not rendering any RPF chains in iptables, I think we should remove this mark combination completely. Will do in a separate PR.

Comment (tomastigera) [reply felix/bpf-gpl/fib.h:215]: https://github.com/projectcalico/calico/pull/6085

File: felix/bpf-gpl/conntrack.h [hunk 9]
```diff
@@ -219,11 +226,11 @@ static CALI_BPF_INLINE int calico_ct_v4_create_tracking(struct ct_create_ctx *ct
...
 				break;
 			}
+		}
 
-			if (i == PSNAT_RETRIES) {
-				CALI_INFO("Source collision unresolved 0x%x:%d\n",
-						bpf_htonl(ip_src), ct_value.orig_sport);
-			}
+		if (i == PSNAT_RETRIES) {
+			CALI_INFO("Source collision unresolved 0x%x:%d\n",
```
Comment (fasaxc) [inline felix/bpf-gpl/conntrack.h:232]: Should we drop in this case?

File: felix/bpf-gpl/tc.c [hunk 10]
```diff
@@ -732,6 +760,21 @@ static CALI_BPF_INLINE struct fwd calico_tc_skb_accepted(struct cali_tc_ctx *ctx
...
+		if (CALI_F_TO_HOST && CALI_F_NAT_IF) {
+			ct_ctx_nat.flags |= CALI_CT_FLAG_VIA_NAT_IF;
+		}
+		if (state->flags & CALI_ST_HOST_PSNAT) {
+			ct_ctx_nat.flags |= CALI_CT_FLAG_HOST_PSNAT;
+		}
+		/* Mark connections that were routed via bpfnatout, but had CT miss at
+		 * HEP. That is because of SNAT happened between bpfnatout and here.
+		 * Returning packets on such a connection must go back via natbpfout
+		 * without a short-circuit to revers the service NAT.
```
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:772]: Typo 
```suggestion
		 * without a short-circuit to reverse the service NAT.
```

**Review by fasaxc — CHANGES_REQUESTED** [review]
Think you missed some of my comments; maybe because I reviewed commit by commit?

File: felix/bpf-gpl/conntrack.h (hunk 9 above)
Comment (tomastigera) [reply felix/bpf-gpl/conntrack.h:232]: probably yes

File: felix/bpf-gpl/tc.c [hunk 11]
```diff
@@ -1194,6 +1269,73 @@ int calico_tc_skb_send_icmp_replies(struct __sk_buff *skb)
...
+
+	__u16 sport = ctx.state->sport;
+	ctx.state->sport = 0;
+	ctx.state->ct_result = calico_ct_v4_lookup(&ctx);
+	ctx.state->sport = sport;
+	ctx.state->flags |= CALI_ST_HOST_PSNAT;
+
+	switch (ct_result_rc(ctx.state->ct_result.rc)) {
+	case CALI_CT_ESTABLISHED:
+		/* Because we are on the "from host" patch, conntrack gives us
```
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:1308]:
```suggestion
		/* Because we are on the "from host" path, conntrack gives us
```

File: felix/bpf/tc/attach.go [hunk 12]
```diff
@@ -659,7 +662,7 @@ func updateJumpMap(obj *libbpf.Obj, isHost bool, ipv6Enabled bool) error {
 		// in IPv6, we add the prologue program to the jump map, and the first entry is 4.
 		base := -1
 		if ipFamily == "IPv6" {
-			base = 4
+			base = 5
```
Comment (fasaxc) [inline felix/bpf/tc/attach.go:665]: Should we future proof by making the map say 32 in size and setting base=16?

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 13]
```diff
@@ -1567,3 +1721,110 @@ func (m *bpfEndpointManager) getInterfaceIP(ifaceName string) (*net.IP, error) {
...
+		"Name":      update.Name,
+		"Namespace": update.Namespace,
+	}).Info("Service Update")
+
+	ips := make([]string, 0, 2+len(update.ExternalIps))
+	if update.ClusterIp != "" {
+		ips = append(ips, update.ClusterIp)
+	}
+	if update.LoadbalancerIp != "" {
+		ips = append(ips, update.LoadbalancerIp)
```
Comment (fasaxc) [inline felix/dataplane/linux/bpf_ep_mgr.go:1736]: Wonder if we could get conflicts for LB/Ext IPs in some scenarios.  What if the Ext IP is also added to the node itself, it might hit `lo` :thinking:

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 14]
```diff
@@ -1567,3 +1721,110 @@ func (m *bpfEndpointManager) getInterfaceIP(ifaceName string) (*net.IP, error) {
...
+		for _, ip := range ips {
+			if err := m.dp.setRoute(ip); err != nil {
+				log.WithError(err).Errorf("Failed to set route to %s via bpfnatin.", ip)
+				errored = true
+			}
+		}
+
+		known := m.services[svc]
+
+		for _, old := range known {
```
Comment (fasaxc) [inline felix/dataplane/linux/bpf_ep_mgr.go:1766]: Wonder if we can see services with overlapping IPs (especially ext IPs) for some transient window.  I suspect we can't rule it out, in which case I think we need to do something to handle it; otherwise I think you'll end up with no route after one of the overlapping service is removed

File: felix/dataplane/linux/bpf_ep_mgr_test.go [hunk 15]
```diff
@@ -122,6 +134,41 @@ func (m *mockDataplane) setAndReturn(vari **polprog.Rules, key string) func() *p
...
+func (m *mockDataplane) delRoute(ip string) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	defer func() { m.shouldErr = false }()
+
+	if m.shouldErr {
+		return fmt.Errorf("delRoute error")
+	}
+
+	delete(m.routes, ip)
```
Comment (fasaxc) [inline felix/dataplane/linux/bpf_ep_mgr_test.go:160]: Error if the route doesn't exist?  (I think that's what the real version does)

File: felix/fv/bpf_test.go [hunk 16]
```diff
@@ -744,6 +752,24 @@ func describeBPFTests(opts ...bpfTestOpt) bool {
 
 			err := infra.AddDefaultDeny()
 			Expect(err).NotTo(HaveOccurred())
+
+			if !testOpts.connTimeEnabled {
```
Comment (fasaxc) [inline felix/fv/bpf_test.go:756]: Any  tests that can have their `testOpts.connTimeEnabled` condition revoked now?

Comment (tomastigera) [reply felix/fv/bpf_test.go:756]: This sets environment prerequisites. It is not necessary if felix sets it on its own. I moved it to felix itself. So when we enable BPF, felix sets all=0 heps=2.

File: felix/bpf/tc/attach.go (hunk 12 above)
Comment (tomastigera) [reply felix/bpf/tc/attach.go:665]: we will resolve this in ipv6 work

**Comment by tomastigera** [comment]
Ooops I did some other work on this branch, had to remove it :man_shrugging:

**Review by fasaxc — APPROVED** [review]
